plonk-core/src/proof_system/proof.rs

```diff
@@ -46,73 +43,80 @@ use rand_core::OsRng;
+    PC: HomomorphicCommitment<F>,
 {
     /// Commitment to the witness polynomial for the left wires.
-    pub(crate) a_comm: Commitment<E>,
+    pub(crate) a_comm: PC::Commitment,
 
     /// Commitment to the witness polynomial for the right wires.
-    pub(crate) b_comm: Commitment<E>,
+    pub(crate) b_comm: PC::Commitment,
 
     /// Commitment to the witness polynomial for the output wires.
-    pub(crate) c_comm: Commitment<E>,
+    pub(crate) c_comm: PC::Commitment,
 
     /// Commitment to the witness polynomial for the fourth wires.
-    pub(crate) d_comm: Commitment<E>,
+    pub(crate) d_comm: PC::Commitment,
 
     /// Commitment to the permutation polynomial.
-    pub(crate) z_comm: Commitment<E>,
+    pub(crate) z_comm: PC::Commitment,
 
     /// Commitment to the quotient polynomial.
-    pub(crate) t_1_comm: Commitment<E>,
+    pub(crate) t_1_comm: PC::Commitment,
 
     /// Commitment to the quotient polynomial.
-    pub(crate) t_2_comm: Commitment<E>,
+    pub(crate) t_2_comm: PC::Commitment,
 
     /// Commitment to the quotient polynomial.
-    pub(crate) t_3_comm: Commitment<E>,
+    pub(crate) t_3_comm: PC::Commitment,
 
     /// Commitment to the quotient polynomial.
-    pub(crate) t_4_comm: Commitment<E>,
+    pub(crate) t_4_comm: PC::Commitment,
+
+    /// Batch opening of the commitments
+    pub aw_opening: PC::Proof,
+
+    /// Batch openings of the shifted commitments
+    pub saw_opening: PC::Proof,
 
     /// Commitment to the opening proof polynomial.
-    pub(crate) w_z_comm: Commitment<E>,
+    //pub(crate) w_z_comm: PC::Commitment,
 
     /// Commitment to the shifted opening proof polynomial.
-    pub(crate) w_zw_comm: Commitment<E>,
+    //pub(crate) w_zw_comm: PC::Commitment,
 
     /// Subset of all of the evaluations added to the proof.
-    pub(crate) evaluations: ProofEvaluations<E::Fr>,
-
-    /// Type Parameter Marker
-    pub(crate) __: PhantomData<P>,
+    pub(crate) evaluations: ProofEvaluations<F>,
 }
 
-impl<E, P> Proof<E, P>
+impl<F, PC> Proof<F, PC>
 where
-    E: PairingEngine,
-    P: TEModelParameters<BaseField = E::Fr>,
+    F: FftField + PrimeField,
+    PC: HomomorphicCommitment<F>,
 {
     /// Performs the verification of a [`Proof`] returning a boolean result.
-    pub(crate) fn verify(
+    pub(crate) fn verify<P>(
         &self,
-        plonk_verifier_key: &PlonkVerifierKey<E, P>,
-        transcript: &mut TranscriptWrapper<E>,
-        verifier_key: &VerifierKey<E>,
-        pub_inputs: &[E::Fr],
-    ) -> Result<(), Error> {
+        plonk_verifier_key: &PlonkVerifierKey<F, PC>,
+        transcript: &mut Transcript,
+        verifier_key: &PC::VerifierKey,
+        pub_inputs: &[F],
+    ) -> Result<(), Error>
+    where
+        P: TEModelParameters<BaseField = F>,
+    {
         let domain =
-            GeneralEvaluationDomain::<E::Fr>::new(plonk_verifier_key.n)
-                .unwrap();
+            GeneralEvaluationDomain::<F>::new(plonk_verifier_key.n).unwrap();
```

Comment:
You should not need to unwrap now that you have the `From` impl

Reply:
We should do something like:
```.ok_or(Error::InvalidEvalDomainSize {
                log_size_of_group: plonk_verifier_key.n.trailing_zeros(),
                adacity: <F as FftField>::FftParams::TWO_ADACITY,
            })?```
            instead of `unwrap()` but I'm having some issues implementing this.
 I could just remove the parameters from `InvalidEvalDomainSize` or something quick, or leave it as `unwrap` and open another issue/PR for this case (since it's rather independent)

Reply:
Sure. Don't want to extend this PR opened more. I've added an issue for it so that we can discuss better the error handling here.